lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -650,6 +652,79 @@ void CBoostedTreeTest::testIntegerRegressor() {
     CPPUNIT_ASSERT(modelRSquared > 0.99);
 }
 
+void CBoostedTreeTest::testMissingData() {
+
+    // Test censoring, i.e. data missing is correlated with the target variable.
+
+    std::size_t rows{1000};
+    std::size_t cols{4};
+    test::CRandomNumbers rng;
+
+    auto frame = core::makeMainStorageDataFrame(cols).first;
+
+    frame->categoricalColumns({false, false, false, false});
+    for (std::size_t i = 0; i < rows - 4; ++i) {
+        frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+            TDoubleVec regressors;
+            rng.generateUniformSamples(0.0, 10.0, cols - 1, regressors);
+            double target{0.0};
+            for (auto regressor : regressors) {
+                *(column++) = regressor > 9.0 ? core::CDataFrame::valueOfMissing() : regressor;
+                target += regressor;
+            }
+            *column = target;
+        });
+    }
+    frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+        *(column++) = core::CDataFrame::valueOfMissing();
+        *(column++) = 2.0;
+        *(column++) = 6.0;
+        *column = core::CDataFrame::valueOfMissing();
+    });
+    frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+        *(column++) = 2.0;
+        *(column++) = core::CDataFrame::valueOfMissing();
+        *(column++) = 6.0;
+        *column = core::CDataFrame::valueOfMissing();
+    });
+    frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+        *(column++) = 2.0;
+        *(column++) = 6.0;
+        *(column++) = core::CDataFrame::valueOfMissing();
+        *column = core::CDataFrame::valueOfMissing();
+    });
+    frame->writeRow([&](core::CDataFrame::TFloatVecItr column, std::int32_t&) {
+        *(column++) = core::CDataFrame::valueOfMissing();
+        *(column++) = 3.0;
+        *(column++) = core::CDataFrame::valueOfMissing();
+        *column = core::CDataFrame::valueOfMissing();
+    });
+    frame->finishWritingRows();
+
+    auto regression = maths::CBoostedTreeFactory::constructFromParameters(
+                          1, std::make_unique<maths::boosted_tree::CMse>())
+                          .buildFor(*frame, cols - 1);
+
+    regression->train();
+    regression->predict();
+
+    TDoubleVec expectedPredictions{17.5, 17.5, 17.5, 22.0};
```

Comment:
We censor for regressors in the range [9, 10] and then add on 1 * "regressor value" to the target. Since we draw regressors uniformly at random we have the additive term is E[X | X > 9], i.e. 9.5 since distribution is uniform. I'll add a comment in the test.

Reply:
Added a comment in [11fe50c](https://github.com/elastic/ml-cpp/pull/625/commits/11fe50c61d2238a81cdc1a29655b12dd580d1c21).